lib/interpreters/split_clustering.rb

```diff
@@ -129,6 +120,302 @@ def self.generate_split_vrps(service_vrp, _job = nil, block = nil)
+        split_vrp: nil,
+      }
+      transfer_empty_vehicles, service_vrp[:split_solve_data][:current_vehicles] =
+        service_vrp[:split_solve_data][:current_vehicles].partition{ |vehicle|
+          service_vrp[:split_solve_data][:service_vehicle_assignments][vehicle.id].empty?
+        }
+      service_vrp[:split_solve_data][:transferred_vehicles].concat(transfer_empty_vehicles)
+      transfer_empty_vehicles.each{ |v| service_vrp[:split_solve_data][:service_vehicle_assignments].delete(v.id) }
+      service_vrp[:split_solve_data][:split_vrp] = create_fake_split_vrp(service_vrp[:split_solve_data])
+
+      # Then split the services into left-and-right groups by
+      # using the service_vehicle_assignments information
+      # (don't generate any sub-VRP yet)
+      new_split_solve_core(service_vrp, job, &block) # self-recursive method
+    ensure
+      service_vrp[:vrp] = service_vrp[:split_solve_data][:original_vrp]
+      service_vrp[:vrp].services.concat empties_or_fills
+      log '<-- new_split_solve (clustering by max_split)'
+    end
+
+    # self-recursive method
+    def self.new_split_solve_core(service_vrp, job = nil, &block)
+      split_level = service_vrp[:split_level]
+      ss_data = service_vrp[:split_solve_data]
+
+      if split_solve_candidate?(service_vrp) # if it still needs splitting
+        enum_current_vehicles = ss_data[:current_vehicles].select # Enumerator to keep a local copy of current_vehicles
+        current_vehicle_limit = ss_data[:current_vehicle_limit]
+
+        # SPLIT current_vehicles list (by-vehicle-centroids) to create two "sides"
+        sides = split_balanced_kmeans(
+          { vrp: create_fake_sub_split_vrp(ss_data) }, 2,
+          cut_symbol: :duration, restarts: 3, build_sub_vrps: false, basic_split: true, group_points: false
+        ).sort_by!{ |side| [side.size, side.sum(&:visits_number)] }.reverse!
+        sides.collect!{ |side| enum_current_vehicles.select{ |v| side.any?{ |s| s.id == v.id } } }
+
+        log 'There should be exactly two clusters in new_split_solve_core!', level: :warn unless sides.size == 2 && sides.none?(&:empty?)
+
+        split_service_counts = sides.collect{ |current_vehicles|
+          current_vehicles.sum{ |v| ss_data[:service_vehicle_assignments][v.id].size }
+        }
+        log "--> new_split_solve_core level: #{split_level} | split services #{split_service_counts.sum}->#{split_service_counts}, vehicles #{enum_current_vehicles.size}->#{sides.collect(&:size)}"
+
+        # RECURSIVELY CALL new_split_solve_core AND MERGE RESULTS
+        results = sides.collect.with_index{ |side, index|
+          service_vrp[:split_level] = split_level + 1
+
+          ss_data[:current_vehicles] = side
+
+          vehicle_limit_ratio = current_vehicle_limit.to_f * side.size / enum_current_vehicles.size
+          # Warning: round does not work if there is an even "half" split
+          ss_data[:current_vehicle_limit] == current_vehicle_limit &&
```

Comment:
What is the purpose of this line as there is no affectation ?

Reply:
`==` performs a comparison and returns a boolean
`&& (index.zero? ? vehicle_limit_ratio.ceil : vehicle_limit_ratio.floor)` is only performed if the previous assertion is true

`vehicle_limit_ratio.ceil` or `vehicle_limit_ratio.floor` are never assigned ?

Reply:
I got your question now! `==` is a bug, it should have been `=` 
(I did not see the mistake, even after your question and writing an answer :facepalm:) 

So it is as follows normally
```ruby
          ss_data[:current_vehicle_limit] = current_vehicle_limit &&
            (index.zero? ? vehicle_limit_ratio.ceil : vehicle_limit_ratio.floor)
```